geomstats/special_orthogonal_group.py

```diff
@@ -86,72 +103,71 @@ def regularize(self, rot_vec):
         :returns self.regularized_rot_vec: 3d vector with: 0 < norm < pi
         """
         assert self.belongs(rot_vec)
-
-        rot_vec = np.array(rot_vec)
         angle = np.linalg.norm(rot_vec)
+        regularized_rot_vec = rot_vec
 
-        self.regularized_rot_vec = rot_vec
         if angle != 0:
             k = np.floor(angle / (2 * np.pi) + .5)
-            self.regularized_rot_vec = (1. - 2. * np.pi * k / angle) * rot_vec
+            regularized_rot_vec = (1. - 2. * np.pi * k / angle) * rot_vec
 
-        return self.regularized_rot_vec
+        return regularized_rot_vec
 
     def rotation_vector_from_matrix(self, rot_mat, epsilon=EPSILON):
         """
         Convert rotation matrix to rotation vector
         (axis-angle representation).
 
+        Get the angle through the trace of the rotation matrix:
+        The eigenvalues are:
+        1, cos(angle) + i sin(angle), cos(angle) - i sin(angle)
+        so that: trace = 1 + 2 cos(angle), -1 <= trace <= 3
+
+        Get the rotation vector through the formula:
+        S_r = angle / ( 2 * sin(angle) ) (R - R^T)
+
+        For the edge case where the angle is close to pi,
+        the formulation is derived by going from rotation matrix to unit
+        quaternion to axis-angle:
+         r = angle * v / |v|, where (w, v) is a unit quaternion.
+
         :param rot_mat: 3x3 rotation matrix
-        :returns rot_vec: 3d rotation vector
+        :return rot_vec: 3d rotation vector
         """
-        assert rot_mat.shape == (self.n, self.n)
+        assert rot_mat.shape == (3, 3)
 
-        # -- Get angle, angle of the rotation
-        # Note: trace(rotation_matrix) = 2cos(angle) + 1
-        cos_angle = np.clip((np.trace(rot_mat) - 1) * 0.5, -1, 1)
+        rot_mat = closest_rotation_matrix(rot_mat)
+
+        trace = np.trace(rot_mat)
+        cos_angle = .5 * (trace - 1)
+        cos_angle = np.clip(cos_angle, -1, 1)
         angle = np.arccos(cos_angle)
 
-        # -- Edge case: angle close to 0
+        rot_vec = vector_from_skew_matrix(rot_mat - rot_mat.transpose())
+
         if angle < epsilon:
```

Comment:
nit: as I see these, I'm thinking that maybe having a function if IsClose(angle, 0) or if IsClose(angle, pi) with Isclose being something like return np.abs(a-b) < epsilon would be nice?